build-system/compile/pre-closure-babel.js

```diff
@@ -69,48 +66,67 @@ function getFilesToTransform() {
 }
 
 /**
- * Apply babel transforms prior to closure compiler pass and return the path
+ * Apply babel transforms prior to closure compiler pass, store the transformed
+ * file in an output directory (used by closure compiler), and return the path
  * of the transformed file.
  *
  * When a source file is transformed for the first time, it is written to a
- * cache directory from where it is retrieved every subsequent time without
- * invoking babel.
+ * persistent transform cache from where it is retrieved every subsequent time
+ * without invoking babel. A change to the file contents or to the invocation
+ * arguments will invalidate the cached result and re-transform the file.
  *
  * @param {string} file
  * @param {string} outputFilename
  * @param {!Object} options
- * @return {string}
+ * @return {Promise<string>}
  */
 async function preClosureBabel(file, outputFilename, options) {
-  if (!cacheDir) {
-    cacheDir = tempy.directory();
+  if (!outputDir) {
+    outputDir = tempy.directory();
+  }
+  if (!transformCache) {
+    transformCache = new TransformCache('.pre-closure-cache', '.js');
   }
   if (!filesToTransform) {
     filesToTransform = getFilesToTransform();
   }
-  const transformedFile = path.join(cacheDir, file);
-  if (fs.existsSync(transformedFile)) {
-    return transformedFile;
-  }
+  const transformedFile = path.join(outputDir, file);
   if (!filesToTransform.includes(file)) {
-    await fs.copy(file, transformedFile);
+    if (!(await fs.exists(transformedFile))) {
+      await fs.copy(file, transformedFile);
+    }
     return transformedFile;
   }
-  debug(CompilationLifecycles['pre-babel'], file);
-  const babelOptions = babel.loadOptions({
-    caller: {name: 'pre-closure'},
-    filename: file,
-    sourceFileName: path.relative(process.cwd(), file),
-  });
   try {
-    const {code} = await babel.transformFileAsync(file, babelOptions);
-    await fs.outputFile(transformedFile, Buffer.from(code, 'utf-8'));
-    debug(CompilationLifecycles['pre-closure'], transformedFile);
+    debug(CompilationLifecycles['pre-babel'], file);
+    const babelOptions =
+      babel.loadOptions({caller: {name: 'pre-closure'}}) || {};
+    const optionsHash = md5(
+      JSON.stringify({babelOptions, argv: process.argv.slice(2)})
+    );
+    const {contents, hash} = await batchedRead(file, optionsHash);
```

Comment:
Not _at all_ necessary for this PR since this is just me musing:

I think the ideal `TransformCache` abstraction might also accept a `transformer`  and `getCacheKey` fn. Then it could do all the reading/transforming (instead of duplicating it at each consumer). We could then also make `batchedRead`/`md5` private.

Reply:
Interesting idea, I like it! I'll try tackling this in a separate PR so the change can be tested in isolation.